java/daikon/dcomp/DCInstrument24.java

```diff
@@ -0,0 +1,4916 @@
+          }
+          // Recurse and check the super_class.
+          classnameStack.push(this_class);
+          this_class = super_class;
+        }
+      }
+
+      // Even if we have not detected that JUnit is active, any class that
+      // contains a method with a RuntimeVisibleAnnotation of org/junit/Test
+      // needs to be marked as a JUnit test class. (Daikon issue #536)
+
+      if (!junit_test_class) {
+        // need to check for JUnit Test annotation on a method
+        searchloop:
+        for (MethodModel mm : classModel.methods()) {
+          for (Attribute<?> attribute : mm.attributes()) {
+            if (attribute instanceof RuntimeVisibleAnnotationsAttribute rvaa) {
+              if (debugJunitAnalysis) {
+                System.out.printf("attribute: %s%n", attribute);
+              }
+              for (final Annotation item : rvaa.annotations()) {
+                String description = item.className().stringValue();
+                if (debugJunitAnalysis) {
+                  System.out.printf("item: %s%n", description);
+                }
+                if (description.endsWith("org/junit/Test;") // JUnit 4
+                    || description.endsWith("org/junit/jupiter/api/Test;") // JUnit 5
+                ) {
+                  junit_test_class = true;
+                  junitTestClasses.add(classname);
+                  break searchloop;
+                }
+              }
+            }
+          }
+        }
+      }
+
+      if (junit_test_class) {
+        debugInstrument.log("JUnit test class: %s%n", classname);
+      } else {
+        debugInstrument.log("Not a JUnit test class: %s%n", classname);
+      }
+    }
+
+    classInfo.isJunitTestClass = junit_test_class;
+
+    instrumentAllMethods(classModel, classBuilder, classInfo);
+
+    // Have all top-level classes implement our interface
+    if (classGen.getSuperclassName().equals("java.lang.Object")) {
+      @SuppressWarnings("signature:assignment")
+      @MethodDescriptor String objectArgReturnBoolean = "(Ljava/lang/Object;)Z";
+      // Add equals method if it doesn't already exist. This ensures
+      // that an instrumented version, equals(Object, DCompMarker),
+      // will be created in this class.
+      MethodModel eq = classGen.containsMethod("equals", objectArgReturnBoolean);
+      if (eq == null) {
+        debugInstrument.log("Added equals method%n");
+        add_equals_method(classBuilder, classGen, classInfo);
+      }
+
+      // Add DCompInstrumented interface and the required
+      // equals_dcomp_instrumented method.
+      add_dcomp_interface(classBuilder, classGen, classInfo);
+    }
+
+    // Output the list of interfaces.
+    classBuilder.withInterfaces(classGen.getInterfaceList());
+
+    // Copy all other ClassElements to output class unchanged.
+    for (ClassElement ce : classModel) {
+      debugInstrument.log("ClassElement: %s%n", ce);
+      switch (ce) {
+        case MethodModel mm -> {}
```

Comment:
Above, there was a loop over all methods, adding instrumentation and adding them to `classBuilder`.  Why is that separate loop necessary?  Would it be possible to operate on each method here, one at a time?  Or is the bulk processing of them all together essential?